[perl] Add annotator README.md in perl/ (#581)
bystrogenomics/bystro: perl/README.md

> # Bystro High Dimensional Genetics Annotator Documentation

Reviewer: What exactly do you mean by high dimensional? WGS or a lot of annotations or both? Does the average geneticist understand high dimensional in general?

Reviewer: I think 'genomics' instead of 'genetics' would also make more sense if you're trying to express that it is 'high dimensional' if you mean that the genetics are the high dimension.
Bystro Genomics Annotator?

Author: I can rename to genomics, but high dimensional indicates big data. Here it is very appropriate, because dimensionality refers to the 'p' in n * p, and we can output a very large p due to the performance of the annotator.

> # Bystro High Dimensional Genomic Variant and Sample Annotator Documentation

Reviewer: This kind of sounds like it annotates samples instead of variants. Maybe it would be helpful to have a quick chat about this?

## What is the Bystro Annotator?

> The Bystro Annotator is the fastest and most comprehensive data curation and labeling library in the world for genetic data. It take 1 or more VCF ([Variant Call Format](https://samtools.github.io/hts-specs/VCFv4.2.pdf)) or SNP ([PEMapper/PECaller](https://www.pnas.org/doi/full/10.1073/pnas.1618065114)) files as input, and outputs a cleaned and thoroughly labeled (annotated) representation of your data, along with a genotype dosage matrix in the [Arrow Feather V2/IPC format](https://arrow.apache.org/docs/python/feather.html), and a set of statistics on the data.

Reviewer: ```suggestion
The Bystro Annotator is the fastest and most comprehensive data curation and labeling library in the world for genetic data. It takes 1 or more VCF ([Variant Call Format](https://samtools.github.io/hts-specs/VCFv4.2.pdf)) or SNP ([PEMapper/PECaller](https://www.pnas.org/doi/full/10.1073/pnas.1618065114)) files as input, and outputs a cleaned and thoroughly labeled (annotated) representation of your data, along with a genotype dosage matrix in the [Arrow Feather V2/IPC format](https://arrow.apache.org/docs/python/feather.html), and a set of statistics on the data.
```

Bystro Annotator processes both variant and sample genotypes. It is capable of processing millions of samples and billions of mutations on commodity hardware such as a laptop or a workstation. It is roughly **100,000** times faster than [Variant Effect Predictor](https://www.ensembl.org/info/docs/tools/vep/index.html) (VEP), **100** times faster than [Annovar](https://doc-openbio.readthedocs.io/projects/annovar/en/latest/), and **50** times faster than [Illumina Connected Annotations](https://developer.illumina.com/illumina-connected-annotations), all while outputting more annotations than any of these tools. What takes VEP years to do, Bystro can do in minutes to hours, all without requiring multiple servers.

> Bystro's performance isn't just about speed, it's also about comprehensiveness. For instance, because of its performance, Bystro can afford to provide complete annotations from gnomad v4.1, for all populations and subpopulations, from the exomes, genomes, and joint datasets, genome-wide. This is intersecting terabytes of data over the entire genome with each variant, all done in microseconds per variant. **No other tool can do this**.

Reviewer: Can you describe the number of fields? Also, might be helpful to hear if other annotators use a lower level of gnomad, a tiny fraction of gnomad, or can't use gnomad at all to get more perspective. It's hard to get a sense of how far forward we are without those kinds of baselines.

## Running Your First Annotation

See the [INSTALL.md#configuring-the-bystro-annotator](./INSTALL.md#configuring-the-bystro-annotator) section for instructions on how to configure the Bystro Annotator

```sh
bystro-annotate.pl --config ~/bystro/config/hg38.yml --threads 32 --input gnomad.genomes.v4.0.sites.chr22.vcf.bgz --output test/my_annotation --compress gz
```

> The above command will annotate the `gnomad.genomes.v4.0.sites.chr22.vcf.bgz` file with the hg38 database, using 32 threads, and output the results to `test`, and will use `my_annotation` as the prefix for output files.

Reviewer: Do we have a link to this gnomad chr22 set so they can follow along with the tutorial?

Bystro's performance isn't just about speed, it's also about comprehensiveness. In statistics "dimensionality" refers to how many parameters/covariates/descriptions a model has. Imagine that we are modeling a genetic variant to try to understand how it impacts a disease. The parameters in the model are the annotations we have for that variant. Bystro can output thousands of annotations for each variant, and can do so for millions of variants in seconds. This is why we call it "high dimensional".

> For exampole, Bystro can afford to provide complete annotations from gnomad v4.1, for all gnomAD populations and subpopulations, from the exomes, genomes, and joint datasets, genome-wide. This means intersecting terabytes of data over the entire genome with each individual variant, all done in microseconds per variant. **No other tool can do this**.

Reviewer: ```suggestion
For example, Bystro can afford to provide complete annotations from gnomad v4.1, for all gnomAD populations and subpopulations, from the exomes, genomes, and joint datasets, genome-wide. This means intersecting terabytes of data over the entire genome with each individual variant, all done in microseconds per variant. **No other tool can do this**.
```

> For exampole, Bystro can afford to provide complete annotations from gnomad v4.1, for all gnomAD populations and subpopulations, from the exomes, genomes, and joint datasets, genome-wide. This means intersecting terabytes of data over the entire genome with each individual variant, all done in microseconds per variant. **No other tool can do this**.

Reviewer: Recommend to be a little more specific than 'this' at the end. Because the last thing you talked about was how fast it was, it sounds like the other tools can still handle terabytes of data but it just takes longer. What is the limit of the rest of the tools in terms of comprehensiveness of annotations?

Author: It's a bit of marketing, but broadly no tool can do any of that

Reviewer: How about adding 'any of'? I do think that it might be impactful to add the limits of other annotations in terms of how much they can annotate if it's specific number of fields or amount of data (versus terabytes).
```suggestion
For example, Bystro can afford to provide complete annotations from gnomad v4.1, for all gnomAD populations and subpopulations, from the exomes, genomes, and joint datasets, genome-wide. This means intersecting terabytes of data over the entire genome with each individual variant, all done in microseconds per variant. **No other tool can do any of this**.
```

The result of this command will be:

```sh
Created completion file
{
   "error" : null,
   "totalProgress" : 8599234,
   "totalSkipped" : 0,
   "results" : {
      "header" : "my_annotation.annotation.header.json",
      "sampleList" : "my_annotation.sample_list",
      "annotation" : "my_annotation.annotation.tsv.gz",
      "dosageMatrixOutPath" : "my_annotation.dosage.feather",
      "config" : "hg38.yml",
      "log" : "my_annotation.annotation.log.txt",
      "statistics" : {
         "qc" : "my_annotation.statistics.qc.tsv",
         "json" : "my_annotation.statistics.json",
         "tab" : "my_annotation.statistics.tsv"
      }
   }
}
```

Explanation of the output:

- `my_annotation.annotation.header.json`: The header of the annotated dataset

- `my_annotation.sample_list`: The list of samples in the annotated dataset

- `my_annotation.annotation.tsv.gz`: A block gzipped TSV file with one row per variant and one column per annotation. Can be decompressed with `bgzip` or any program compatible with the gzip format, like `gzip` and `pigz`.

> - `my_annotation.dosage.feather`: The dosage matrix file, where the first column is the `locus` column in the format "chr:pos:ref:alt", and columns following that are sample columns, with the dosage of the variant for that sample (0 for homozygous reference, 1 for 1 copy of the alternate allele, 2 for 2, and so on). -1 indicates missing genotypes. The dosage is the expected number of alternate alleles, given the genotype. This is useful for downstream analyses like imputation, or for calculating polygenic risk scores

Reviewer: How about for extra copies, are they still 2? I don't remember, are trisomic sites generally handled?

Author: we talk about that

with the dosage of the variant for that sample (0 for homozygous reference, 1 for 1 copy of the alternate allele, 2 for 2, and so on)

Author: I'll add a note that we do not assume diploid

> - `my_annotation.dosage.feather`: The dosage matrix file, where the first column is the `locus` column in the format "chr:pos:ref:alt", and columns following that are sample columns, with the dosage of the variant for that sample (0 for homozygous reference, 1 for 1 copy of the alternate allele, 2 for 2, and so on). -1 indicates missing genotypes. The dosage is the expected number of alternate alleles, given the genotype. This is useful for downstream analyses like imputation, or for calculating polygenic risk scores

Reviewer: ```suggestion
- `my_annotation.dosage.feather`: The dosage matrix file, where the first column is the `locus` column in the format "chr:pos:ref:alt", and columns following that are sample columns, with the dosage of the variant for that sample (0 for homozygous reference, 1 for 1 copy of the alternate allele, 2 for 2, and so on). -1 indicates missing genotypes. The dosage is the expected number of alternate alleles, given the genotype. This is useful for any downstream analyses like association studies or calculating polygenic risk scores.
```

Reviewer: I'm suggesting changing out imputation because I think of imputation more upstream than downstream generally even if it can be down downstream as well. Could be any other analysis as an example if you want to change it out.

Author: just removed imputation

The Bystro Annotator is the fastest and most comprehensive data curation and labeling library in the world for genetic data. It takes 1 or more VCF ([Variant Call Format](https://samtools.github.io/hts-specs/VCFv4.2.pdf)) or SNP ([PEMapper/PECaller](https://www.pnas.org/doi/full/10.1073/pnas.1618065114)) files as input, and outputs a cleaned and thoroughly labeled (annotated) representation of the data, along with a genotype dosage matrix in the [Arrow Feather V2/IPC format](https://arrow.apache.org/docs/python/feather.html), as well as a set of statistics that describe sample-level characteristics of the data.

> Bystro Annotator annotates variants as well as sample genotypes. It is capable of processing millions of samples and billions of mutations on commodity hardware such as a laptop or a workstation. It is roughly **100,000** times faster than [Variant Effect Predictor](https://www.ensembl.org/info/docs/tools/vep/index.html) (VEP), **100** times faster than [Annovar](https://doc-openbio.readthedocs.io/projects/annovar/en/latest/), and **50** times faster than [Illumina Connected Annotations](https://developer.illumina.com/illumina-connected-annotations), all while outputting more annotations than any of these tools. What takes VEP years to do, Bystro can do in minutes to hours, all without requiring multiple servers.

Reviewer: ```suggestion
Bystro Annotator annotates variants as well as sample genotypes. It is capable of processing millions of samples and billions of mutations on commodity hardware such as a laptop or a workstation. It is roughly **100,000** times faster than [Variant Effect Predictor](https://www.ensembl.org/info/docs/tools/vep/index.html) (VEP), **100** times faster than [Annovar](https://doc-openbio.readthedocs.io/projects/annovar/en/latest/), and **50** times faster than [Illumina Connected Annotations](https://developer.illumina.com/illumina-connected-annotations), all while outputting more annotations than any of these tools. What would take VEP years to do, Bystro can do in minutes to hours, all without requiring multiple servers.
```

- The `inputRef` (reference base) in Bystro's annotation outputs is always exactly 1 base long

- The `pos' (position) in Bystro's annotation outputs is always the first affected base, except in the case of insertions, where it is the base before the insertion, since the insertion by definition is between two reference bases

- It is possible to represent all multiallelic site using a single reference base, a single position, and a list of alleles

The Bystro Genotype Dosage Matrix, is a columnar dataset, generated for every collection of VCFs submitted. Its first column is the `locus`, which is `chr:pos:ref:alt`. Every column after that is labeled by the sample name, and contains a -1 for missing genotypes, 0 for reference, 1 for a single alternate allele, 2 for 2 alternate alleles (homozygous in a diploid organism), and so on. It can be used for many things, such as to calculate polygenic risk scores.

### Comparing the VCF and Bystro Variant Representations

> Below we'll demonstrate the Bystro Annotation and Dosage Matrix outputs, using some examples. You may run these examples, if you have Bystro installed, or at least the Bystro VCF Preprocessor. The Bystro VCF Preprocessor, which is a tool that takes a VCF file and outputs a Bystro-annotator compatible TSV file, which is then used by the Perl Bystro Annotator to add annotations to the VCF file from the Bystro Annotation Database. You can install it by running `go install github.com/bystrogenomics/bystro-vcf@2.2.3`.

Reviewer: ```suggestion
Below we'll demonstrate the Bystro Annotation and Dosage Matrix outputs, using some examples. You may run these examples, if you have Bystro or the Bystro VCF Preprocessor installed. The Bystro VCF Preprocessor, a tool that takes a VCF file and outputs a Bystro-annotator compatible TSV file, is used by the Perl Bystro Annotator to add annotations to the VCF file from the Bystro Annotation Database. You can install it by running `go install github.com/bystrogenomics/bystro-vcf@2.2.3`.
```

Reviewer: I'm still a little confused by the Bystro VCF preprocessor sentence. Maybe split it into two like below? Is the final output at the end of this sentence still a VCF format or an annotation file?

The Bystro VCF Preprocessor is a tool that takes a VCF file and outputs a Bystro-annotator compatible TSV file. It is used by the Perl Bystro Annotator to add annotations to the VCF file from the Bystro Annotation Database.

Author: I updated the section, is it ok now?

> Please note that we are not showing the full Bystro Annotator outputs, which are far too extensive to easily display here.

Reviewer: Recommend to just give numbers (just used example numbers below) so that it doesn't sound like it's impossible to look at or think about, but it is large enough that it's clearly impractical to show everything.

```suggestion
Please note that we are not showing the full Bystro Annotator outputs, just the first 20 fields out of the 150 fields that we annotate
```

Author: done

```
cat ~/bystro/perl/example_vcf.tsv | bystro-vcf --keepId --emptyField "NA" --keepPos
```

> If you also want to output the **Bystro Genotype Dosage Matrix**, at a small performance hit, you can run:

Reviewer: Not sure it's necessary to mention a performance hit if it's small? I would assume asking something to do more would take longer but if it's 'small' then it doesn't seem worth drawing attention to it.

Author: The performance comparisons were without this, so I need to make a reference to it.

| chr20 | 4       | MNP          | C        | A   | 2    | NA              | 0              | NA00003     | 0.5          | NA00001      | 0.333       | 2   | 4   | 0.5       | 3       | EXAMPLE_MISSING_MNP         |
| chr20 | 5       | MNP          | C        | A   | 2    | NA              | 0              | NA00003     | 0.5          | NA00001      | 0.333       | 2   | 4   | 0.5       | 3       | EXAMPLE_MISSING_MNP         |

Expected Bystro Genotype Dosage Matrix Output:

```python
import pandas as pd
df = pd.read_feather('example_vcf_dosage_matrix.feather')
print(df)
```

| locus              | NA0001 | NA0002 | NA0003 |
| ------------------ | ------ | ------ | ------ |
| chr20:1:A:T        | 1      | 0      | 0      |
| chr20:1110696:A:G  | 1      | 1      | 0      |
| chr20:1110696:A:T  | 1      | 1      | 2      |
| chr20:1:A:+C       | 0      | 0      | 1      |
| chr20:1:A:+CC      | 1      | 1      | 2      |
| chr20:1234568:T:-3 | 1      | 0      | 2      |
| chr20:1234568:T:+A | 0      | 1      | 0      |
| chr20:3:C:A        | -1     | 0      | 2      |
| chr20:4:C:A        | -1     | 0      | 2      |
| chr20:5:C:A        | -1     | 0      | 2      |

> - Note that missing genotypes are reprsented as -1 in the genotype dosage matrix output

Reviewer: ```suggestion
- Note that missing genotypes are represented as -1 in the genotype dosage matrix output
```

> - If a sample's genotype contains any missing genotypes, the sample is considered missing for the site

Reviewer: I feel like this sentence is a little confusing. It feels like it's saying if a sample is missing at a site, the sample will be missing at a site. Let me know if it's supposed to mean something else.

Author: Should be clearer now?

#### Explanation for SIMPLE_SNP

VCF Representation:

| CHROM | POS | ID         | REF | ALT | QUAL | FILTER | INFO | FORMAT      | NA00001        | NA00002        | NA00003        |
| ----- | --- | ---------- | --- | --- | ---- | ------ | ---- | ----------- | -------------- | -------------- | -------------- |
| 20    | 1   | SIMPLE_SNP | A   | T   | 50   | PASS   | .    | GT:GQ:DP:HQ | 0/1:54:7:56,60 | 0/0:48:4:51,51 | 0/0:48:4:51,51 |

> Bystro Representation:

Reviewer: Recommend some kind of way to show that this is just a representative example, something like either one below.
```suggestion
Bystro Representation (VCF-related fields):
```
```suggestion
Bystro Representation (first 17/150 fields):
```

Reviewer: Maybe this makes more sense at line 144

Author: I think it is more than 150, 220 for hg38 is what I remember. 

I guess my question is, why is that important?

Author: The point of this section is to describe how we handle variants compared to the standard VCF format. The number of annotations isn't that important. I just wanted to mention that this isn't the full output (`see below for the description` of what is in the full output of the default hg38 and hg19 databases), lest they be confused that they only see 17 columns.

Reviewer: So that people get a general sense that this is just a small representative example, though I do think it makes more sense to put this up at the description at line 121

Author: I would also think "VCF-related" may add confusion, because really all of the annotations are VCF related (the sites come from the VCF file in these examples), and at the same time, only 2 of them are (vcfPos and id are specific to VCF files; SNP files don't have them; all other fields pertain to both). Let's just keep it as is?

Author: > So that people get a general sense that this is just a small representative example, though I do think it makes more sense to put this up at the description at line 121

What do you mean by that? This whole section is about how we handle variants vs the VCF spec.

Reviewer: Yep I'm ok with keeping the tables as they are. Maybe the top of the fields description is a good spot for the total number? If it's not an impressive sounding metric or number, we can also just skip it in general.

Author: Sure, I can add it to that section

Author: added

| chrom | pos | type | inputRef | alt | trTv | heterozygotes | heterozygosity | homozygotes | homozygosity | missingGenos | missingness | ac  | an  | sampleMaf | vcfPos | id         |
| ----- | --- | ---- | -------- | --- | ---- | ------------- | -------------- | ----------- | ------------ | ------------ | ----------- | --- | --- | --------- | ------ | ---------- |
| chr20 | 1   | SNP  | A        | T   | 2    | NA00001       | 0.333          | NA          | 0            | NA           | 0           | 1   | 6   | 0.167     | 1      | SIMPLE_SNP |

Bystro Genotype Dosage Matrix Output:

| locus       | NA0001 | NA0002 | NA0003 |
| ----------- | ------ | ------ | ------ |
| chr20:1:A:T | 1      | 0      | 0      |

> The Bystro and VCF formats for simple, well-normalized SNPs are the same. In addition to the position, variant type, reference, and alternate, Bystro's VCF preprocessor (bystro-vcf) also outputs whether a variant is a transition (1), transversion (2) or neither (0), descriptive information about the genotypes, including which samples are heterozyogtes, homozygotes, or missing genotypes, vcfPos (which describes the original position in the VCF file, pre-normalization), and the VCF ID. Meanwhile the genotype dosage matrix output shows the number of alternate alleles for each sample at each variant.

Reviewer: ```suggestion
The Bystro and VCF formats for simple, well-normalized SNPs are the same. In addition to the position, variant type, reference, and alternate, Bystro's VCF preprocessor (bystro-vcf) also outputs whether a variant is a transition (1), transversion (2) or neither (0), descriptive information about the genotypes, including which samples are heterozygotes, homozygotes, or missing genotypes, vcfPos (which describes the original position in the VCF file, pre-normalization), and the VCF ID. Meanwhile the genotype dosage matrix output shows the number of alternate alleles for each sample at each variant.
```

# Bystro High Dimensional Genomic Annotator Documentation

## What is the Bystro Annotator?

> In order to make use of genetic data, which is composed of variants (colloquially known as mutations) and the individuals (samples) that have those variants, we need to clean that data and then carefully describe those variants with as much accurate information as possible. This information can then be used for qualitative and quantitative studies of the variants/samples. These descriptions are known as labels, model parameters, features, or "annotations", depending on your field. These activites are collectively called data curation and data labeling, but in genetics we call them qc and annotation.

Reviewer: ```suggestion
In order to make use of genetic data, which is composed of variants (colloquially known as mutations) and the individuals (samples) that have those variants, we need to clean that data and then carefully describe those variants with as much accurate information as possible. This information can then be used for qualitative and quantitative studies of the variants/samples. These descriptions are known as labels, model parameters, features, or "annotations", depending on your field. These activites are collectively called data curation and data labeling, but in genetics we call them quality control and annotation.
```

The Bystro Annotator is the fastest and most comprehensive data curation and labeling library in the world for genetic data. It takes 1 or more VCF ([Variant Call Format](https://samtools.github.io/hts-specs/VCFv4.2.pdf)) or SNP ([PEMapper/PECaller](https://www.pnas.org/doi/full/10.1073/pnas.1618065114)) files as input, and outputs a cleaned and thoroughly labeled (annotated) representation of the data, along with a genotype dosage matrix in the [Arrow Feather V2/IPC format](https://arrow.apache.org/docs/python/feather.html), as well as a set of statistics that describe sample-level characteristics of the data.

> Bystro Annotator annotates variants as well as sample genotypes. It is capable of processing millions of samples and billions of mutations on commodity hardware such as a laptop or a workstation. It is roughly **100,000** times faster than [Variant Effect Predictor](https://www.ensembl.org/info/docs/tools/vep/index.html) (VEP), **100** times faster than [Annovar](https://doc-openbio.readthedocs.io/projects/annovar/en/latest/), and **50** times faster than [Illumina Connected Annotations](https://developer.illumina.com/illumina-connected-annotations), all while outputting more annotations than any of these tools. What would take VEP years to do, Bystro can do in minutes to hours, all without requiring multiple servers.

Reviewer: VEP and Annovar links seem to no longer be active

Author: fixed

The Bystro Genotype Dosage Matrix, is a columnar dataset, generated for every collection of VCFs submitted. Its first column is the `locus`, which is `chr:pos:ref:alt`. Every column after that is labeled by the sample name, and contains a -1 for missing genotypes, 0 for reference, 1 for a single alternate allele, 2 for 2 alternate alleles (homozygous in a diploid organism), and so on. It can be used for many things, such as to calculate polygenic risk scores.

### Comparing the VCF and Bystro Variant Representations

Below we'll demonstrate how the Bystro Annotator handles different kinds of variants, using some examples. We'll do these demonstartions using the Bystro VCF preprocessor, which is a Go program used by the Bystro Annotator to convert a VCF into a partially annotated tab-separated output. The Bystro VCF preprocessor is installed with Bystro Annotator (see the [INSTALL.md](./INSTALL.md) file for instructions on how to install Bystro Annotator). If you don't have Bystro Annotator installed, you can still run the examples as long as you install the Bystro VCF preprocessor by running `go install github.com/bystrogenomics/bystro-vcf@2.2.3`.

> Please note that we are not showing the full Bystro Annotator output described below. We're showing just the first 17 columns of the output, which are the most important for understanding the variant representation and sample genotype handling.

Reviewer: Maybe here add that it's 17 out of (total number) instead of below.

Author: The total number depends on the database used

Reviewer: What's the max number we currently have?

Author: I'd have to check, it's something like 220
